### ISSUE TITLE
Runtime errors

### ISSUE THREAD

**Issue by jurezabkar** [ddemidov/ev3dev-lang-python#9]
I wrote the following python script that goes through all the sensors and checks all the modes for each sensor:

```python
#!/usr/bin/python

import sys, argparse, time
from ev3dev import *

cs = color_sensor();          assert cs.connected
ts = touch_sensor();          assert ts.connected
gs = gyro_sensor();           assert gs.connected
us = ultrasonic_sensor();   assert us.connected

sensors = [us, cs, gs, ts]

for s in sensors:
    print s.driver_name, "on port", s.port_name
    print "bin_data_format",s.bin_data_format
    for m in s.modes:
        print "\t",m
        s.mode = m
        print "\tnumber of values", s.num_values
        print "\tunits:", s.units
        print "\tvalue", s.value()
    print s.mode, s.modes
    print "\tnumber of values", s.num_values
    print "\tunits:", s.units
    print "\tvalue", s.value()
    print "================================================="
```

The script exits with RuntimeError: No such device:
```
root@ev3dev:~# python test_sensors.py 
lego-ev3-us on port in1
bin_data_format s8
	US-DC-CM
	number of values 1
	units: cm
	value 0
	US-DC-IN
	number of values 1
	units: inch
	value 0
	US-DIST-CM
Traceback (most recent call last):
  File "test_sensors.py", line 18, in <module>
    s.mode = m
RuntimeError: No such device
```
The second run of the script gives similar output a bit earlier:
```
root@ev3dev:~# python test_sensors.py 
lego-ev3-us on port in1
bin_data_format u8
	US-DC-CM
Traceback (most recent call last):
  File "test_sensors.py", line 18, in <module>
    s.mode = m
RuntimeError: No such device
```
Trying to debug, I wrote simpler code, assigning 'US-DC-CM' to us.mode, since that seems to be a problem:
```python
#!/usr/bin/python

import sys, argparse, time
from ev3dev import *

cs     = color_sensor();        assert cs.connected
ts     = touch_sensor();        assert ts.connected
gs     = gyro_sensor();         assert gs.connected
us     = ultrasonic_sensor();   assert us.connected

sensors = [us, cs, gs, ts]
us.mode = 'US-DC-CM'
```
This time, the error is different:
```
root@ev3dev:~# python test_sensors.py 
Traceback (most recent call last):
  File "test_sensors.py", line 12, in <module>
    us.mode = 'US-DC-CM'
RuntimeError: Invalid argument
```
Trying to assign a different mode - 'US-SI-CM' instead of 'US-DC-CM' again results in another type of RuntimeError: Connection timed out:
```
root@ev3dev:~# python test_sensors.py 
Traceback (most recent call last):
  File "test_sensors.py", line 12, in <module>
    us.mode = 'US-SI-CM'
RuntimeError: Connection timed out
```
The computer and the the robot are connected to a local WiFi router and the connection is still on! I can immediately run the **original script** again and it would crash every time in line 18 but not necessarily for the same sensor or sensor mode. In most of the cases, I get *RuntimeError: Connection timed out* but other two errors mentioned above can also happen, most often immediately after I restart the EV3. Besides restarting, I also tried different routers, but as I expected that had no influence.

**Comment by ddemidov** [comment]
What happens when you do
```
$ cat US-DIST-CM > /sys/class/lego-sensor/sensor*/mode
$ cat /sys/class/lego-sensor/sensor*/value0
```
where `sensor*` corresponds to the ultrasonic sensor?

**Comment by jurezabkar** [comment]
root@ev3dev:/sys/class/lego-sensor/sensor5# cat
/sys/class/lego-sensor/sensor5/modes
US-DIST-CM US-DIST-IN US-LISTEN US-SI-CM US-SI-IN US-DC-CM US-DC-IN

root@ev3dev:/sys/class/lego-sensor/sensor5# cat US-DIST-CM >
/sys/class/lego-sensor/sensor5/mode
cat: US-DIST-CM: No such file or directory

root@ev3dev:/sys/class/lego-sensor/sensor5# cat
/sys/class/lego-sensor/sensor5/value0
132

**Comment by ddemidov** [comment]
Sorry, the first cat should be echo of course.

On Tue, Aug 25, 2015, 18:09 jurezabkar <notifications@github.com> wrote:

> root@ev3dev:/sys/class/lego-sensor/sensor5# cat
> /sys/class/lego-sensor/sensor5/modes
> US-DIST-CM US-DIST-IN US-LISTEN US-SI-CM US-SI-IN US-DC-CM US-DC-IN
>
> root@ev3dev:/sys/class/lego-sensor/sensor5# cat US-DIST-CM >
> /sys/class/lego-sensor/sensor5/mode
> cat: US-DIST-CM: No such file or directory
>
> root@ev3dev:/sys/class/lego-sensor/sensor5# cat
> /sys/class/lego-sensor/sensor5/value0
> 132
>
>
> On Tue, Aug 25, 2015 at 3:30 PM, Denis Demidov <notifications@github.com>
> wrote:
>
> > What happens when you do
> >
> > $ cat US-DIST-CM > /sys/class/lego-sensor/sensor*/mode
> > $ cat /sys/class/lego-sensor/sensor*/value0
> >
> > where sensor* corresponds to the ultrasonic sensor?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/ddemidov/ev3dev-lang-python/issues/9#issuecomment-134585854
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ddemidov/ev3dev-lang-python/issues/9#issuecomment-134617335>
> .
>
-- 

Cheers,
Denis

**Comment by jurezabkar** [comment]
That seems to work well:

root@ev3dev:/sys/class/lego-sensor/sensor5# echo US-DIST-CM >
/sys/class/lego-sensor/sensor5/mode
root@ev3dev:/sys/class/lego-sensor/sensor5# cat
/sys/class/lego-sensor/sensor5/value0
326

**Comment by ddemidov** [comment]
I can reproduce the error with an infrared sensor (don't have an ultrasonic one), both with the appropiately modified python script and with the equivalent bash script:
```bash
#!/bin/bash

for sen in /sys/class/lego-sensor/*; do
    cat ${sen}/driver_name
    cat ${sen}/port_name
    cat ${sen}/bin_data_format
    for m in $(cat ${sen}/modes); do
        echo ${m}
        echo ${m} > ${sen}/mode
        cat ${sen}/num_values
        cat ${sen}/units
        cat ${sen}/value0
    done
    cat ${sen}/mode
    cat ${sen}/modes
    cat ${sen}/num_values
    cat ${sen}/units
    cat ${sen}/value0
done
```

Here is the output of the above script (I have touch, color, and infrared sensors connected in that order):
```
$ ./test.sh 
lego-ev3-touch
in4
u8
TOUCH
./test.sh: line 9: echo: write error: Invalid argument
1

0
TOUCH
TOUCH
1

0
lego-ev3-color
in1
s8
COL-REFLECT
1
pct
0
COL-AMBIENT
1
pct
0
COL-COLOR
1
col
0
REF-RAW
2

643
RGB-RAW
3

0
COL-CAL
4

0
COL-CAL
COL-REFLECT COL-AMBIENT COL-COLOR REF-RAW RGB-RAW COL-CAL
4

0
lego-ev3-ir
in2
s16
IR-PROX
1
pct
100
IR-SEEK
8
pct
0
IR-REMOTE
4
btn
0
IR-REM-A
1

384
IR-S-ALT
./test.sh: line 9: echo: write error: Connection timed out
1

384
IR-CAL
./test.sh: line 9: echo: write error: Connection timed out
1

384
IR-REM-A
IR-PROX IR-SEEK IR-REMOTE IR-REM-A IR-S-ALT IR-CAL
1

384
```

[Documentation](http://www.ev3dev.org/docs/sensors/lego-ev3-infrared-sensor/) states that `IR-S-ALT` mode is not useable, so I guess this is expected. @dlech, I don't see anything like it in the [ultrasonic sensor](http://www.ev3dev.org/docs/sensors/lego-ev3-ultrasonic-sensor/) [pages](http://www.ev3dev.org/docs/sensors/lego-nxt-ultrasonic-sensor/). Is this a known problem?

**Comment by dlech** [comment]
The LEGO EV3 Ultrasonic sensor does not have any defective modes that I know of. It sounds like a race condition. Maybe some fresh eyes will help if you would like to have a look. The driver code is [here](https://github.com/ev3dev/lego-linux-drivers/blob/master/sensors/ev3_uart_sensor_ld.c#L293). 

The "No such device" error comes from `-ENODEV`, so it means that the `tty` has disappeared (highly unlikely) or that `port->synced` or `port->info_done` is zero. I'm thinking we should move that check down below the two other checks below it.

When writing to the `mode` attribute, it should block until the mode has actually switched, but this may not be happening.

Or it could be that the microprocessor in the sensor can't handle switching so many times so quickly and it is resetting itself.

Bottom line is, if you add a delay after calling `s.mode = m`, it should work a bit better.

**Comment by ddemidov** [comment]
> The driver code is here.

@dlech, I can not see anything standing out there (not that I have any experience with kernel hacking). It seems though that the only way one could get a `Connection timeout` is when [number of retries](https://github.com/ev3dev/lego-linux-drivers/blob/31bcb5e4df3b0a61da3863021377807e09793344/sensors/ev3_uart_sensor_ld.c#L299) is not enough [here](https://github.com/ev3dev/lego-linux-drivers/blob/31bcb5e4df3b0a61da3863021377807e09793344/sensors/ev3_uart_sensor_ld.c#L321-L331). Is it possible that 10 is too small a number? How would adding a delay *after* calling `s.mode = m` help here? Or is it cumulative effect from the previous calls?

@jurezabkar, could you please test if adding `time.sleep(100)` after `s.mode = m` solves the problem for you?

> or that port->synced or port->info_done is zero. I'm thinking we should move that check down below the two other checks below it

@dlech, can you even touch the `port` struct before the check is done? I could be wrong, but It looks like some kind of port sanity check that has to be done before the others.

**Comment by jurezabkar** [comment]
> @dlech, I can not see anything standing out there (not that I have any experience with kernel hacking). It seems though that the only way one could get a Connection >timeout is when number of retries is not enough here. Is it possible that 10 is too small a number? How would adding a delay after calling s.mode = m help here? Or is it >cumulative effect from the previous calls?



> @jurezabkar, could you please test if adding time.sleep(100) after s.mode = m solves the problem for you?

I ran a huge set of experiments this morning and found out that:
 - setting a delay does not solve the problem (tried different times,
not exactly 100 secs, but only from 0.1 to 5 secs)
 - setting a mode never fails with color sensor (I tried connecting it
to all 4 ports and it always works); it fails with touch, gyro and
ultrasonic sensor.
 - during the first few experiments after switching on the brick,
"RuntimeError: No such device" and "RuntimeError: Invalid argument"
appeared; after awhile, I only get "RuntimeError: Connection timed
out"; cannot reproduce this, but seems strange.
 - some modes cause significantly less problems than others, e.g.
gs.mode = 'GYRO-G&A' always works, while other modes for gs would fail
in the following code:

from ev3dev import *

cs     = color_sensor();        assert cs.connected
ts     = touch_sensor();        assert ts.connected
gs     = gyro_sensor();         assert gs.connected
us     = ultrasonic_sensor();   assert us.connected

sensors = [cs, gs, us, ts]
us.mode = 'US-DIST-CM'
gs.mode = 'GYRO-G&A'
us.mode = 'US-DIST-CM'
gs.mode = 'GYRO-G&A'
us.mode = 'US-DIST-CM'
gs.mode = 'GYRO-G&A'
us.mode = 'US-DIST-CM'
gs.mode = 'GYRO-G&A'
us.mode = 'US-DC-CM'
us.mode = 'US-DC-IN'

**Comment by ddemidov** [comment]
> not exactly 100 secs, but only from 0.1 to 5 secs)

Sorry about that, I forgot that `time.sleep()` takes seconds, not milliseconds :)

> it fails with touch

I think touch sensor does not support mode setting.

**Comment by ddemidov** [comment]
There is a [note](http://www.ev3dev.org/docs/sensors/lego-ev3-ultrasonic-sensor/#fn:dc-mode) on the ultrasonic sensor page linked to `US-DC*` modes:

> NOTE: If you write the mode too frequently (e.g. every 100msec), the sensor will sometimes lock up and writing to the mode attribute will return an error. A delay of 250msec between each write to the mode attribute seems sufficient to keep the sensor from locking up.

Could this be related?

**Comment by jurezabkar** [comment]
> I think touch sensor does not support mode setting.

I thought it does, like in EV3 software (http://www.stemhq.org/touch-sensor)

**Comment by jurezabkar** [comment]
> There is a note on the ultrasonic sensor page linked to US-DC* modes:
>
> NOTE: If you write the mode too frequently (e.g. every 100msec), the sensor will sometimes lock up and writing to the mode attribute will return an error. A delay of 250msec between each write to the mode attribute seems sufficient to keep the sensor from locking up.
>
> Could this be related?

I don't think so - see below that US-DC* modes worked fine; I used
time delay of 1s in the script that results in:

root@ev3dev:~# python test_sensors.py
US-DC-CM
US-DC-IN
US-DIST-CM
US-DIST-IN
US-LISTEN
US-SI-CM
Traceback (most recent call last):
  File "test_sensors.py", line 19, in <module>
    s.mode = m
RuntimeError: Connection timed out

I also tried USB connection instead of WiFi but as expected, it
doesn't solve the problem.

**Comment by ddemidov** [comment]
@dlech, should we open an issue on evdev/ev3dev? This does not seem to be binding-specific.

**Comment by dlech** [comment]
Yes, that would be good. Be sure to include a script to reproduce the problem.

**Comment by ddemidov** [comment]
@jurezabkar, can you please attach the output of the following script on your system here or in [the new issue](https://github.com/ev3dev/ev3dev/issues/new) in the core repository? I guess it should be enough to reproduce the problem:
```bash
#!/bin/bash

for sen in /sys/class/lego-sensor/*; do
    cat ${sen}/driver_name
    cat ${sen}/port_name
    for m in $(cat ${sen}/modes); do
        echo ${m}
        echo ${m} > ${sen}/mode
    done
done
```

**Comment by ddemidov** [comment]
Replaced with ev3dev/ev3dev#385.
